### PR TITLE
test-bot: revise Step output transcoding and XML character filtering

### LINKED ISSUE #16: Jenkins runs fail on invalid XML: "Invalid byte 2 of 4-byte UTF-8 sequence"
Some Jenkins build runs will fail completely without output like this. No test results or artifacts will be available.

```
Failed to read test report file /Users/brew/Jenkins/prs/version/yosemite/brew-test-bot.xml
org.dom4j.DocumentException: Invalid byte 2 of 4-byte UTF-8 sequence. Nested exception: Invalid byte 2 of 4-byte UTF-8 sequence.
    at org.dom4j.io.SAXReader.read(SAXReader.java:484)
    at org.dom4j.io.SAXReader.read(SAXReader.java:264)
    at hudson.tasks.junit.SuiteResult.parse(SuiteResult.java:123)
    at hudson.tasks.junit.TestResult.parse(TestResult.java:286)
    at hudson.tasks.junit.TestResult.parsePossiblyEmpty(TestResult.java:232)
    at hudson.tasks.junit.TestResult.parse(TestResult.java:167)
```

The test output is all included in an XML file, which includes the output of some of the steps. That output, from commands we call, may not be valid XML CDATA, or even valid UTF-8, since we're calling external commands.

The `test-bot` command [attempts to filter out invalid characters](https://github.com/Homebrew/brew/blob/ac8088304603a95e2ae4d187016a50c36c622605/Library/Homebrew/cmd/test-bot.rb#L945-L946) but it's apparently not being thorough enough. We need more robust filtering or escaping.

I think the approach in https://github.com/Homebrew/legacy-homebrew/pull/50092 was pretty solid, but needed different syntax, like so.

```
output = step.output.gsub(/[^\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u{10000}-\u{10FFFF}]/, "")
```

And it might be better to replace filtered-out characters with a Unicode "replacement character" placeholder instead of just eliding. Then there'd still be some indication that there was other output there.

###  References 

* Blocking https://github.com/Homebrew/legacy-homebrew/pull/49971
* https://github.com/Homebrew/legacy-homebrew/pull/50092 was an attempt to fix, but didn't work out
 * a6b918d6c5cd8a05494002d288f15d9b7b30f9df – original
 * 60055ba8901dfe2844c323101d2e425ce800cf2e – fixup
 * 161a1e286ec238cc9449a1d3e40362ce49e207bc – reverted to old code
* Caused breakage in https://github.com/Linuxbrew/linuxbrew/pull/1037#event-612849650

### PR DESCRIPTION
**Pull request by apjanke**
- [X] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/.github/CONTRIBUTING.md) document?
- [X] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [X] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031). *Ans*: nope; couldn't figure out how to put test-bot in a test harness or isolate this code.
- [X] Have you successfully ran `brew tests` with your changes locally?

##  Description

Fixes #16.

This takes care of the occasional "Invalid XML" or "Invalid UTF-8" errors we see on Jenkins, where it cannot read

This change alters `test-bot` to:
  * Buffer the steps' output as binary, and only convert it to UTF-8 once collected, to avoid transcoding problems with multibyte characters.
  * Use a more robust regexp to filter out characters that are not allowed in XML. Instead of excluding some known-bad characters, it lists the known-valid character ranges and excludes everything else.
  * Replace filtered-out characters with the `xFFFD` Unicode replacement character instead of just deleting them, so it's obvious when there was junk in a test step's output.
  * Replace the single-byte `read(1)` with `readpartial` and a larger buffer, for performance.
  * Handle an edge case in the truncation code's size calculation.

### CODE REVIEW

**Comment by apjanke** [comment]
Reopening for revision: the original submission broke under Ruby 1.8.7, because it doesn't support the `\u` escapes. And Ruby 1.9 doesn't support `\u{...}`.

Jenkins builds on Mavericks were failing with stuff like this.

```
==> git clean -ffdx
HEAD is now at ceed663 test-bot: revise Step output transcoding and XML character filtering
Removing Cellar/
Removing Library/Homebrew/test/.bundle/
Removing Library/Homebrew/test/fs_leak_log
Removing Library/Homebrew/test/vendor/
Error: invalid regular expression: /[^    

 -uD7FFuE000-uFFFDu{10000}-u{10FFFF}]/
Please report this bug:
```

**Comment by mikemcquaid** [comment]
FWIW we may be able to just use the Regex on Ruby 2.0 (or for now at least) given the previous regex seemed to handle more situations there and Ruby <2.0 is on the way out for us.

**Comment by apjanke** [comment]
Problem is that when expressed as a regex literal, it gets parsed in the compilation stage or whatever, not when the line is executed. So even if you stick it in a `if ... /new-regex/ ... else ... /old-regex/ ... end` block, Ruby 1.8.7 still throws an "invalid regex" error, even if it never executes the line in question.

I'm trying a workaround now where I invert the regex to `/[\x00-\x08\x0B\x0C\x0E-\x1F\uD800-\uDFFF\uFFFE\uFFFF]/` (selecting the invalid XML characters), which is still syntactically valid for Ruby 1.8 even if it wouldn't match the right characters, so it can still be used inside an `if / else` block.

**Comment by mikemcquaid** [comment]
Ah yes, I see.

**Comment by apjanke** [comment]
I've updated the PR with the new 1.8-safe code. Will try pushing it tomorrow: it's about bed time for me, and GitHub is being pretty slow at the moment.

File: Library/Homebrew/cmd/test-bot.rb [hunk 1]
```diff
@@ -944,11 +953,21 @@ def test_bot
 
           if step.has_output?
             # Remove invalid XML CData characters from step output.
-            output = step.output.delete("\000\a\b\e\f\x2\x1f")
+            valid_xml_chars = ""
+            if ruby_has_encoding?
+              # This is the regex for valid XML chars, but only works in Ruby 2.0+
+              # /[]\x09\x0A\x0D\x20-\uD7FF\uE000-\uFFFD\u{10000}-\u{10FFFF}]/
```
Comment (xu-cheng) [inline Library/Homebrew/cmd/test-bot.rb]: `/[]`->`/[^` ?

File: Library/Homebrew/cmd/test-bot.rb [hunk 2]
```diff
@@ -944,11 +953,21 @@ def test_bot
...
+              # For 1.9 compatibility, use the inverse of that, which staus under \u10000
```
Comment (UniqMartin) [inline Library/Homebrew/cmd/test-bot.rb]: `staus` -> `stays`?

File: Library/Homebrew/cmd/test-bot.rb [hunk 3]
```diff
@@ -944,11 +953,21 @@ def test_bot
...
+              invalid_xml_pat = /[\x00-\x08\x0B\x0C\x0E-\x1F\uD800-\uDFFF\uFFFE\uFFFF]/
+              output = step.output.gsub(invalid_xml_pat, "\uFFFD")
+            else
+              output = step.output.delete("\000\a\b\e\f\x2\x1f")
+            end
 
             if output.bytesize > BYTES_IN_1_MEGABYTE
-              output = "truncated output to 1MB:\n" \
-                + output.slice(-BYTES_IN_1_MEGABYTE, BYTES_IN_1_MEGABYTE)
+              slice_start = [0, -BYTES_IN_1_MEGABYTE].max
```
Comment (UniqMartin) [inline Library/Homebrew/cmd/test-bot.rb]: Isn't that always going to return 0? Shouldn't this be `[0, output.length - BYTES_IN_1_MEGABYTE].max`? Does that already account for the case that for a typical UTF-8 string `bytesize > length` and does it make sure not to cut into the middle of some UTF-8 byte sequence or a single code unit?

And more Unicode complications: What if that cuts into the middle of some grapheme cluster? It won't make the UTF-8 invalid, but it will disturb the output at the cut-off point. (Probably details we shouldn't care about.)

**Comment by UniqMartin** [comment]
> - Buffer the steps' output as binary, and only convert it to UTF-8 once collected, to avoid transcoding problems with multibyte characters.

Can this be modified to transcode those bytes that are already valid UTF-8 and leave the remaining few bytes, that could be the beginning of a valid UTF-8 sequence, in the buffer? Given that UTF-8 is a prefix code that should be doable. That said, I'm not sure if it's worth the effort (maybe for another PR), i.e., if there are scenarios that would benefit from having partial output available.

**Comment by apjanke** [comment]
>  Can this be modified to transcode those bytes that are already valid UTF-8 and leave the remaining few bytes, that could be the beginning of a valid UTF-8 sequence, in the buffer?

That's what I mean by "incremental transcoding with remainder support". Certainly possible, but a lot of work. If someone knows a library that can do it, sure, but no way I'm going to write it. (You'd also have to do invalid-sequence replacement as you go, to avoid stalling.)

File: Library/Homebrew/cmd/test-bot.rb (hunk 3 above)
Comment (apjanke) [inline Library/Homebrew/cmd/test-bot.rb]: No, because the code is calculating the slice start and end points in bytes, but the arguments to `slice()` are characters. My adjustment here is to account for that bytesize vs character differences and prevent overruns.

And the fact that `slice()` takes characters prevents it from cutting in to the middle of a UTF-8 byte sequence. (It doesn't account for surrogate pairs, though. That's an edge case, even in terms of validity, and I doubt we'd ever see it.)

Grapheme clusters: yeah, but we're already truncating at an arbitrary point; it's probably going to be in the middle of a word, so the stuff right around the cutoff is probably going to be junk to a human anyway.

Comment (apjanke) [inline Library/Homebrew/cmd/test-bot.rb]: Oh, wait. I misread you and focused on the UTF-8 stuff. You're right, that'll always be 0. Adjusting...

**Comment by apjanke** [comment]
Something's gone wonky with my attempts to add further amended commits to this PR. I re-opened this as a fresh PR #27, including adjustments from recent feedback. Let me know if you have any comments; I'll try pushing again in a couple hours and see if it works. *This* time I'm confident I actually got the "test under ruby 1.8.7" right.
